src/tests/system/tests/test_authentication.py

```diff
@@ -0,0 +1,86 @@
+
+:requirement: offline
+"""
+from __future__ import annotations
+
+import pytest
+from sssd_test_framework.roles.client import Client
+from sssd_test_framework.roles.generic import GenericProvider
+from sssd_test_framework.topology import KnownTopologyGroup
+
+
+@pytest.mark.topology(KnownTopologyGroup.AnyProvider)
+@pytest.mark.parametrize("method", ["su", "ssh"])
+def test_authentication__login(client: Client, provider: GenericProvider, method: str):
+    """
+    :title: ssh/su login
+    :setup:
+        1. Add user to SSSD
+        2. Set password for user
+        3. Start SSSD
+    :steps:
+        1. Authenticate user with correct password
+        2. Authenticate user with incorrect password
+    :expectedresults:
+        1. User is authenticated
+        2. User is not authenticated
+    :customerscenario: False
+    """
+    provider.user("user1").add(password="Secret123")
+
+    client.sssd.start()
+
+    assert client.auth.parametrize(method).password("user1", "Secret123"), "login with correct password failed"
+    assert not client.auth.parametrize(method).password("user1", "NOTSecret123"), "login with wrong password succeeded"
+
+
+@pytest.mark.topology(KnownTopologyGroup.AnyProvider)
+@pytest.mark.parametrize("method", ["su", "ssh"])
+def test_authentication__offline_login(client: Client, provider: GenericProvider, method: str):
+    """
+    :title: Offline ssh/su login
+    :setup:
+        1. Add user to SSSD and set its password
+        2. In SSSD domain change "cache_credentials" and "krb5_store_password_if_offline" to "True"
+        3. In SSSD pam change "offline_credentials_expiration" to "0"
+        4. Start SSSD
+    :steps:
+        1. Authenticate user with wrong password
+        2. Authenticate user with correct password
+        3. Make server offline (by removing firewall rules for LDAP, KDC and Global Catalog ports)
+        4. Bring SSSD offline explicitly
+        5. Offline authentication of user with correct password
+        6. Offline authentication of user with wrong password
+    :expectedresults:
+        1. User is not authenticated
+        2. User is authenticated
+        3. Firewall rules dropped
+        4. SSSD is offline
+        5. Offline authentication is successful
+        6. Offline authentication is not successful
+    :customerscenario: False
+    """
+    user = "user1"
+    correct = "Secret123"
+    wrong = "Wrong123"
+    provider.user(user).add(password=correct)
+
+    client.sssd.domain["cache_credentials"] = "True"
+    client.sssd.domain["krb5_store_password_if_offline"] = "True"
+    client.sssd.pam["offline_credentials_expiration"] = "0"
+    client.sssd.start()
+
+    assert not client.auth.parametrize(method).password(user, wrong), "login with wrong password succeeded"
+    assert client.auth.parametrize(method).password(user, correct), "login with correct password failed"
+
+    # Block KDC, LDAP and Global Catalog ports.
+    provider.firewall.drop([88, 389, 3268])
```

Comment:
636 is not missing since ldaps is not used, you would have to configure SSSD to use it.

You usually want more realistic scenario, inaccessible ports is more realistic to me than blocking outgoing traffic, but of course, both can be done.

Reply:
Also the host can be accessible via multiple IP addresses and interfaces so it would be probably quite hard to achieve that universally.

Reply:
I looked into this. outgoing rules in firewall-cmd are quite problematic. They now have support for "Policy Objects" that can be used to do that or fallback to direct iptables rules.

Blocking outgoing connections to specific IP address would be problematic, but I think achievable - we would need to get the list of all IP addresses that the target host has available. Alternatively, we could block all outgoing connections.

Either was, we need to allow specific ports that should remain open (like DNS for this test case).

I don't really see any benefit here, except we could speed up using REJECT instead of DROP rules so we don't have to wait for timeouts. DROP is the real life scenario though (we have to reach timeout in order to see that a host is not available). Otherwise it is the same, including the need to terminate active connections.

We can implement this if you want, but I think it is out of scope of this pull request and we should move on.

Reply:
In this case the port 639 is not used but we got tests that use it like this one: https://github.com/SSSD/sssd/blob/master/src/tests/multihost/ad/test_adparameters.py#L578 .
For now consider this good to go.